### PR TITLE
SDL-0296 Add ability to show navi streaming

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Use ffmpeg to proxy streaming from original SDL output (socket or
pipe) to webm http streaming on 8085 port

Put streaming from ffmpeg to video tag of the navi applicaiton
Part of https://adc.luxoft.com/jira/browse/FORDTCN-4167

### CODE REVIEW

File: tools/start_server.py [hunk 1]
```diff
@@ -215,13 +221,34 @@ def handle_get_app_manifest_message(params):
 
 	return json.dumps(response_msg)
 
+
+def handle_start_streaming_adapter(params):
+	print("-->Handle start gstreamer adapter\r")
```
Comment (LuxoftAKutsan) [inline tools/start_server.py]:
```suggestion
	print("-->Handle start ffmpeg adapter\r")
```

File: README.md [hunk 2]
```diff
@@ -34,6 +14,31 @@ git submodule update
...
+
+## Start HMI
+A quick guide to installing, configuring, and running HMI.
+
+	1. run `deploy_server.sh`
+	2. run SmartDeviceLinkCore
+	2. run chromium-browser [root_of_cloned_sdl_hmi_repo/index.html]
+
+## Simulating signals for LOW_VOLTAGE feature
+I
```
Comment (AKalinich-Luxoft) [inline README.md]:
```suggestion

```

File: app/controller/InfoController.js [hunk 3]
```diff
@@ -546,6 +546,46 @@ SDL.InfoController = Em.Object.create(
         );
     },
 
+    
+    startStreamingAdapter: function(url) {
```
Comment (AKalinich-Luxoft) [inline app/controller/InfoController.js:554]: @LuxoftAKutsan please add a description for this

File: app/controller/InfoController.js [hunk 4]
```diff
@@ -546,6 +546,46 @@ SDL.InfoController = Em.Object.create(
...
+        let client = FFW.RPCSimpleClient;
+
+        let response_timer = setTimeout(function() {
+          Em.Logger.log('startVideoStreamingAdapter timout');
+          client.unsubscribeFromEvent('StartStreamingAdapter');
+          reject();
+        }, 10000);
+
+        let response_callback = function(params) {
+          Em.Logger.log('StreamingAdapter response');
```
Comment (AKalinich-Luxoft) [inline app/controller/InfoController.js]:
```suggestion
          Em.Logger.log('StartStreamingAdapter response');

```

File: app/controller/InfoController.js [hunk 5]
```diff
@@ -546,6 +546,46 @@ SDL.InfoController = Em.Object.create(
         );
     },
 
+    
+    startStreamingAdapter: function(url) {
+      return new Promise( (resolve, reject) => {
+        let client = FFW.RPCSimpleClient;
+
+        let response_timer = setTimeout(function() {
+          Em.Logger.log('startVideoStreamingAdapter timout');
```
Comment (AKalinich-Luxoft) [inline app/controller/InfoController.js]:
```suggestion
          Em.Logger.log('StartStreamingAdapter timeout');

```

File: app/model/sdl/Abstract/Model.js [hunk 6]
```diff
@@ -663,20 +663,24 @@ SDL.SDLModel = Em.Object.extend({
...
-          playPromise.then(_ => {
-            console.log('Video playback started OK');
-          })
-          .catch(error => {
-            console.log('Video playback start failed: ' + error);
-            SDL.SDLModel.data.naviVideo = null;
-          });
-        }
+        SDL.InfoController.startStreamingAdapter(sdl_stream).then(function(stream_endpoint) {
+          console.log('Adapter promice callback');
```
Comment (AKalinich-Luxoft) [inline app/model/sdl/Abstract/Model.js]:
```suggestion
          console.log('Starting video playback');
```

File: tools/start_server.py [hunk 7]
```diff
@@ -215,13 +221,34 @@ def handle_get_app_manifest_message(params):
...
+	stream_endpoint = "http://localhost:8085"
```
Comment (AKalinich-Luxoft) [inline tools/start_server.py]: @LuxoftAKutsan could you please move port number to a separate constant at line 48?

File: tools/start_server.py [hunk 8]
```diff
@@ -215,13 +221,34 @@ def handle_get_app_manifest_message(params):
...
+	ffmpeg_process = ffmpeg.input(params['url']).output(stream_endpoint, vcodec="vp8", format="webm", listen=1, multiple_requests=1).run_async(pipe_stderr=True) 
+	print("Wait for data from SDL")
+	o = pexpect.fdpexpect.fdspawn(ffmpeg_process.stderr.fileno())
+	o.expect("Input")
+	print("data from SDL available")
```
Comment (AKalinich-Luxoft) [inline tools/start_server.py]:
```suggestion
	print("Data from SDL is available")
```

File: tools/start_server.py [hunk 9]
```diff
@@ -215,13 +221,34 @@ def handle_get_app_manifest_message(params):
 
 	return json.dumps(response_msg)
 
+
+def handle_start_streaming_adapter(params):
+	print("-->Handle start gstreamer adapter\r")
+	stream_endpoint = "http://localhost:8085"
+	ffmpeg_process = ffmpeg.input(params['url']).output(stream_endpoint, vcodec="vp8", format="webm", listen=1, multiple_requests=1).run_async(pipe_stderr=True)
```
Comment (AKalinich-Luxoft) [inline tools/start_server.py:239]: @LuxoftAKutsan suggest to add check that `params` actually contains `url` for performing some basic validation

Comment (LuxoftAKutsan) [reply tools/start_server.py:239]: bba3698

File: tools/start_server.py (hunk 7 above)
Comment (LuxoftAKutsan) [reply tools/start_server.py]: bba3698

File: app/controller/InfoController.js (hunk 3 above)
Comment (LuxoftAKutsan) [reply app/controller/InfoController.js:554]: bba3698

File: tools/start_server.py [hunk 10]
```diff
@@ -37,8 +37,16 @@
...
+
+import ffmpeg
+import threading
+import pexpect.fdpexpect
+
+
 WEBSOCKET_PORT = 8081
 FILESERVER_PORT = 8082
+HTML5_STREAMING_HOST = "http://localhost"
+HTML5_STREAMING_PORT = 8085
```
Comment (YarikMamykin) [inline tools/start_server.py:49]: Is it important to point exactly HTML**5** in variable names?
I think you can remove `HTML5_` prefix in these names.

**Review by YarikMamykin — CHANGES_REQUESTED** [review]
@LuxoftAKutsan by the way, how do we handle some error stuff from `ffmpeg`?
Or it is not actually applicable?

File: tools/start_server.py (hunk 10 above)
Comment (LuxoftAKutsan) [reply tools/start_server.py:49]: Not agree. this streaming is specially for sending it to <video> tag ( HTML5)

Comment (YarikMamykin) [reply tools/start_server.py:49]: Okek then.

**Comment by LuxoftAKutsan** [comment]
>@LuxoftAKutsan by the way, how do we handle some error stuff from ffmpeg?
Or it is not actually applicable?

in case of any error from ffmpeg it will throw it into terminal
We will see python excptin in server script ( it will not crash the script) 
fffmpeg error output and error in js console in browser : Connection refused

**Review by YarikMamykin — APPROVED** [review]

**Pull request merged**